ACPs/99-validatorsetmanager-contract/README.md

```diff
@@ -0,0 +1,264 @@
+    uint64 weight,
+    // The start time of the validation period
+    uint64 startTime,
+    // The end time of the validation period (only ≠ 0 when the period is over)
+    uint64 endTime,
+    // The uptime of the validator during this validation period (only ≠ 0 once the validation period is over)
+    uint64 uptime
+}
+
+/**
+ * The list of message IDs associated with a validator of the Subnet
+ * Each message ID corresponds to a RegisterSubnetValidatorTx
+ * NodeID => messageID[]
+ */
+mapping(bytes32 => bytes32[]) subnetValidatorMessages;
+// The validation corresponding to each message ID (= RegisterSubnetValidatorTx)
+// messageID => SubnetValidation
+mapping(bytes32 => SubnetValidation) subnetValidations;
+```
+
+#### About `validation`s
+
+A `validation` represents the continuous time frame during which a node is part of the validator set and can be composed of multiple periods. A new period starts every time the validator weight changes during the same validation.
+
+Each validation is identified by the `MessageID` corresponding to the Warp message in the `RegisterSubnetValidatorTx` that added the validator.
+
+`subnetValidations` will allow custom “security modules” (e.g. a PoS module) to take appropriate action upon node operator interactions (e.g. rewarding the node operator upon unstake). The `SubnetValidationPeriod` list will enable proper reconstruction of the validation timeline for each validator.
+
+### External functions
+
+The `ValidatorSetManager` contract should implement the following `external` functions.
+
+**Note:** This is pseudo-code that might not be 100% syntactically correct.
+
+```solidity
+// Returns the ID of the Subnet tied to this contract
+function getSubnetID() external view returns (bytes32);
+
+// Returns a Subnet validator information by its NodeID
+function getSubnetValidator(bytes32 nodeID) external view returns (SubnetValidator);
+
+/**
+ * Used to issue a RegisterSubnetValidatorTx
+ * @param nodeID The ID of the node to add to the Subnet
+ * @param weight The weight of the node on the Subnet
+ * @param expiry The time after which this message is invalid
+ * @param signature The Ed25519 signature of [subnetID]+[nodeID]+[blsPublicKey]+[weight]+[timestamp]
+ */
+function registerSubnetValidator(
+    bytes32 nodeID,
+    uint64 weight,
+    uint64 expiry,
+    bytes signature
+) external onlyOwner;
+
+/**
+ * Used to issue a SetSubnetValidatorWeightTx
+ * @param nodeID The ID of the node to modify
+ * @param weight The new weight of the node on the Subnet
+ */
+function setSubnetValidatorWeight(
+    bytes32 nodeID,
+    uint64 weight
+) external onlyOwner;
+
+/**
+ * Called to update the Subnet validator set upon receiving a notification Warp message from the P-Chain
+ * (after the successful addition or removal of a validator)
+ * @param messageIndex The index of the Warp message from the P-Chain
+ */
+function updateSubnetValidatorSet(
+    uint32 messageIndex
+) external;
```

Comment:
```suggestion
) external;

/** Used to issue a SetSubnetValidatorManagerTx
  * @param blockchainID The ID of the chain on which the Subnet validator manager is located
  * @param managerAddress The address of the Subnet validator set manager
  */
function setSubnetValidatorManager(
    bytes32 blockchainID,
    address managerAddress
) external onlyOwner;
```
This function is gonna be needed in case we want to change the validator manager address of the Subnet.

Reply:
`SetSubnetValidatorManagerTx` was removed from ACP-77